keps/sig-node/2000-graceful-node-shutdown/README.md

```diff
@@ -0,0 +1,864 @@
+[acpid](https://wiki.archlinux.org/index.php/Acpid) daemon which uses a variety
+of mechanisms to watch ACPI events (i.e. reading `/proc/acpi/event` or
+`/dev/input/eventX` to react to power button presses). However, in most modern
+linux distros today,
+[systemd-logind](https://www.freedesktop.org/wiki/Software/systemd/logind/)
+has taken over as the main component [reacting to ACPI
+events](https://wiki.archlinux.org/index.php/Power_management#ACPI_events) and
+initiating shutdown of the machine. On a system with
+systemd-logind, for example, a trigger of the power button will
+result in the systemd target
+[poweroff](https://www.freedesktop.org/software/systemd/man/systemd-halt.service.html)
+being run (see
+[HandlePowerKey](https://www.freedesktop.org/software/systemd/man/logind.conf.html),
+which will terminate all the systemd services running on the machine and
+eventually shut it down. However, in the context of kubernetes, systemd is not
+aware of the pods and containers running on the machine and systemd will simply
+kill them as regular linux processes.
+
+### Background on Inhibitors
+
+`systemd-logind` provides the ability for applications to delay shutdown and
+perform some series of actions before the shutdown completes through a
+mechanism called ["Inhibitor
+Locks".](https://www.freedesktop.org/wiki/Software/systemd/inhibit/)
+Applications can request to delay shutdown by taking an inhibitor lock by
+sending messages to logind over dbus. Applications can request up to
+`InhibitDelayMaxSec` (a setting configured in `logind.conf`) for delay based
+locks, which allow applications to receive sleep and shutdown events, and block
+the shutdown from proceeding by `InhibitDelayMaxSec` period to execute some
+critical work prior to shutdown/sleep.
+
+We believe that making use of systemd is a reasonable approach considering
+almost all new popular linux distros are systemd based (RHEL, Google COS,
+Ubuntu, CentOS, Debian, Fedora, Flatcar Linux, see widespread
+[adoption](https://en.wikipedia.org/wiki/Systemd#Adoption)) and [systemd
+183](https://lwn.net/Articles/499480/) (released in 2012) features support for
+inhibitors.
+
+Thanks to @giuseppe for helping with getting systemd inhibitors working!
+
+### Implementation
+
+Introduce a new Kubelet Config setting, `kubeletConfig.ShutdownGracePeriod`,
+defaulting to 0 seconds. Upon kubelet startup,
+
+*   if the setting is greater than 0 seconds
+    *   kubelet will check with dbus current `InhibitDelayMaxSec` to check if
+        `kubeletConfig.ShutdownGracePeriod &lt;=` `InhibitDelayMaxSec`.
+*   if `kubeletConfig.ShutdownGracePeriod` > `InhibitDelayMaxSec`
+    *   Kubelet will attempt to update the InhibitDelayMaxSec setting, by
+        writing a config file to `/etc/systemd/logind.conf.d/kubelet.conf`, and
+        sending a SIGHUP to logind to update the config setting to ensure that
+        the ShutdownGracePeriod from kubelet config is equal to
+        `InhibitDelayMaxSec`.
```

Comment:
Reconfiguring the node through kubelet config is ok with me, but not sure if there are concerns from the community. It is worth mentioning at SIG Node meeting.

Reply:
We discussed this in SIG-Node today, I didn't hear any strong concerns from the community, so I think this should be fine for now.